chips/sam4l/src/adc.rs

```diff
@@ -537,7 +537,12 @@ impl Adc {
                 // becomes: N <= ceil(log_2(f(CLK_CPU)/1500000)) - 2
                 let cpu_frequency = pm::get_system_frequency();
                 let divisor = (cpu_frequency + (1500000 - 1)) / 1500000; // ceiling of division
-                clock_divisor = math::log_base_two(math::closest_power_of_two(divisor)) - 2;
+                clock_divisor = math::log_base_two(math::closest_power_of_two(divisor));
+                if clock_divisor > 2 {
```

Comment:
I'm not sure this is the right fix. The assumption in line 546 is that clock_divisor is signed. That's why it runs the max of (clock_divisor, 0). An alternative fix, that I think would be more appropriate, would be to change line 496 to `let mut clock_divisor: i8;`.

Reply:
You're right that that's the assumption with which the code was originally written. The only issue with using i8 is it requires a lot of casts because `closest_power_of_two()` and `log_base_two()` and `PRESCAL.val()` all expect u32, but I'll go ahead and write it that way.

Reply:
Given that you are producing an unsigned value, I think it's less bug prone to calculate it as an unsigned value and treat it as such, rather than compute a signed value and convert it to unsigned at the end.

Reply:
I agree with @phil-levis.

I think what we actually want is to remove the `cmp::max(clock_divisor, 0)` on line 546, and instead maintain the invariant that we never overflow, by using `checked_sub` (instead of the manual if statement):

```rust
clock_divisor = clock_divisor.checked_sub(2).unwrap_or(0);
```

And really, since it's better style to avoid `mut` variables, and even better style to declare and initialize variables at the same time, go with the slightly more verbose:

```rust
let divisor_pow2 = math::closest_power_of_two(divisor)
let clock_divisor =
    math::log_base_two(divisor_pow2).checked_sub(2).unwrap_or(0);
```

Reply:
Yeah, given that `log_base_two` operates on `u32`s and not signed numbers, the assumption is definitely incorrect. I like Amit's solution above for the problem. We should also remove the `cmp::max...cmp::min` stuff (in both branches).

Although I'm with you on the avoiding `mut` point, the reason it's mutable is because there are two different branches of an if statement each of which needs to set the variable before another instruction after the if statements actually writes it to a register.

Reply:
I pushed a fix that is basically Amit's fix. I removed the use of mut by mutating `cfg_val` directly inside both branches instead of once outside the branch using a mutated `clock_divisor`. However, I left in the cmp::min() that limits the maximum value for `clock_divisor` to 7. 

Currently, the code as-is could never lead to a `clock_divisor` greater than 7 if the maximum system clock speed is assumed to be 48 MHz. However, if the PLL was set to its maximum frequency (240 MHz) and divided down for the CPU, `clock_divisor` would be set to 8 without the check. This scenario is pretty highly unlikely, and would only happen in some use case where multiple peripherals required very specific clock speeds which could only be achieved by dividing down an extremely high frequency clock. 

That being said, I also think there is some value in keeping `cmp::min(x, 7)` because it explicitly shows that the value shouldn't be greater than 7 in case anyone changes this code in the future.

I'm happy to remove the check entirely if people disagree though -- I'm not entirely sure that setting the `clock_divisor` to 7 when it should be 8 per the `divisor` is even correct, but that seems to be the assumption built in originally.

Reply:
The answer is that writing 7 to the 3-bit register is a lot less wrong than writing 8 to it. Still probably won't work, but at least it's not writing to other nearby bits too.

Reply:
Yeah the question is should that just return ReturnCode::FAIL instead, but I think writing a 7 is probably preferable, as for certain requested sampling frequencies you could still get correct operation (I believe)

Reply:
Does the Register interface prevent an overflow/underflow on a field from writing to nearby bits?

Reply:
It does do that. However since it masks that would end up setting the clock divisor to 0 rather than 7.